README.md

```diff
@@ -0,0 +1,123 @@
+
+# Remote Job Executor
+
+## **Introduction**
+This document outlines the design approach for developing a remote job executor that provides functionalities such as job creation, monitoring, control, and secure data transfer. Our primary objective is to design a system that is secure, efficient, and user-friendly.
+
+## **Scope**
+
+In this design, we aim to implement the level 5 requirements. This means that we need to implement:
+
+1. A CLI for communicating with the API server
+2. A library that allows clients to create jobs with resource control and isolation and allows multiple clients to stream a job's output.
+3. A gRPC server with mTLS authentication and a simple authorization scheme.   
+
+
+## **Design Approach**
+
+### **1. Command-Line User Experience (CLI UX)**
+
+The CLI will allow users to start and stop a job, query a job's state and stream their output. Below are some example commands. For job ids we are using the 12 character UUIDs. We are not using sequential numerical ids to prevent malicious users from guessing job ids.   
+
+* **Starting a Job:**  
+  ```
+  $ cli start echo "Hello World" 
+  Started job with ID: 4d5c9fa91db4
+  ```
+
+* **Starting a Job with flags/options:**
+  ```
+  # Per convention -- indicates the end of command options
+  $ cli start -- bash -c "echo Hello World"
+  Started job with ID: a2f3d7865b4e
+  ```
+
+* **Stopping a Job:**  
+  ```
+  $ cli stop 4d5c9fa91db4
+  ```
+
+* **Query Job Status:**
+  ```
+  $ cli status 4d5c9fa91db4
+  ```
+
+* **Viewing Job Output:**  
+  ```
+  $ cli log 4d5c9fa91db4
+  ```
+
+* **Streaming Job Ouput:**
+  ```
+  $ cli log -f 4d5c9fa91db4
+  ```
+
+#### **Future CLI Improvements**
+
+It's worth mentioning that our CLI UX is basic and we would probably want to build something better in production. To improve our user CLI experience we could add:
+
+* **Job Listing**
+  ```
+  $ cli ls
+  ```
+
+* **Job Naming**
+  ```
+  $ cli start -n hello_world -- echo Hello World 
+  ```
+
+For now, we'll leave these two improvements for the future to reduce our implementation scope. 
+
+### **2. Library**
+
+#### **Interface**
+
+The library will expose the following functions as the primary interface:
+
+```go
+type struct Job {
+  id
+  <internal_fields>
+}
+type struct LogReader {
+  <internal_fields>
+}
+
+// Managing jobs and queueing job status
+StartJob(command string, args ...string) (*Job, error)
+func (job *Job) Stop() error
+func (job *Job) GetStatus() error
+
+// Reading job output
+func (job *Job) GetNewLogReader() (*LogReader, error)
+func (reader *LogReader) GetNextLineBlocking()
```

Comment:
Also, what happens if the output from the process is not just text in lines?

Reply:
Great point about blocking forever. Also, I guess I was just dealing with the happy case where the output are lines.

So passing in the context here doesn't seem to me to be the right move because this is supposed to be a generic library. Also, I realized that implementing a blocking read because you have to implement a timeout since there is a chance you never receive a next line. Also, we would have to make sure we temporarily release the log reader read-lock so that we can write to the job log. How about if we just use a non-blocking read and busy wait for now for simplicity? 

To deal with bytes, how about we change this API to :
```
func (reader *LogReader) GetNextStdOutBytesNonBlocking() []byte
func (reader *LogReader) GetNextStdErrBytesNonBlocking() []byte
```

One issue that I see here is that the `cli log` command is a bit more complicated to implement because we have to deal with two stream of bytes instead of a single stream of lines. For comparison, I think that what kubectl here does is to use heuristics to identify whether the stdout/stderr streams are strings and only prints the log output if it is a string output. 

How about, for simplicity, we only deal with the happy case in the `cli log` command where we assume the output is always a string output?

Reply:
Are there any well-known and commonly implemented interfaces we could leverage for reading bytes from the job instead of creating our own API here?

Reply:
> How about if we just use a non-blocking read and busy wait for now for simplicity?
That's a bit too much of a stretch, I think.

> How about, for simplicity, we only deal with the happy case in the cli log command where we assume the output is always a string output?
Even then, what if there's a 50MiB text line?